m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,831 @@
+        Buckets buckets,
+        Duration bucketLowerBound,
+        Duration bucketUpperBound,
+        long samples
+    ) {
+        synchronized (isOpenLock) {
+            if (!isOpen) {
+                LOG.error("Reporter already closed; no more metrics can be flushed");
+                return;
+            }
+            // Append histogram bucket-specific tags
+            int bucketIdLen = String.valueOf(buckets.size()).length();
+            bucketIdLen = Math.max(bucketIdLen, MIN_METRIC_BUCKET_ID_TAG_LENGTH);
+
+            String bucketIdFmt = String.format("%%0%sd", bucketIdLen);
+
+            BucketPair[] bucketPairs = BucketPairImpl.bucketPairs(buckets);
+
+            if (tags == null) {
+                tags = new HashMap<>(2, 1);
+            }
+
+            for (int i = 0; i < bucketPairs.length; i++) {
+                // Look for the first pair with an upper bound greater than or equal
+                // to the given upper bound.
+                if (bucketPairs[i].upperBoundDuration().compareTo(bucketUpperBound) >= 0) {
+                    String idTagValue = String.format(bucketIdFmt, i);
+
+                    tags.put(bucketIdTagName, idTagValue);
+                    tags.put(bucketTagName,
+                        String.format("%s-%s",
+                            durationBucketString(bucketPairs[i].lowerBoundDuration()),
+                            durationBucketString(bucketPairs[i].upperBoundDuration())
+                        )
+                    );
+
+                    break;
+                }
+            }
+
+            reportCounterInternal(name, tags, samples);
+        }
+    }
+
+    // Relies on the calling function to provide guarantees of the reporter being open
+    private void reportCounterInternal(String name, Map<String, String> tags, long value) {
+        CountValue countValue = new CountValue();
+        countValue.setI64Value(value);
+
+        MetricValue metricValue = new MetricValue();
+        metricValue.setCount(countValue);
+
+        Metric metric = newMetric(name, tags, metricValue);
+
+        queueSizedMetric(new SizedMetric(metric, calculateSize(metric)));
+    }
+
+    private Metric newMetric(String name, Map<String, String> tags, MetricValue metricValue) {
+        Metric metric = new Metric(name);
+        metric.setTags(toMetricTagSet(tags));
+        metric.setTimestamp(System.currentTimeMillis() * Duration.NANOS_PER_MILLI);
+        metric.setMetricValue(metricValue);
+
+        return metric;
+    }
+
+    private void queueSizedMetric(SizedMetric sizedMetric) {
+        try {
+            metricQueue.put(sizedMetric);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+
+            LOG.warn(String.format("Interrupted queueing metric: %s", sizedMetric.getMetric().getName()));
+        }
+    }
+
+    private class Processor implements Runnable {
+        List<Metric> metrics = new ArrayList<>(freeBytes / 10);
```

Comment:
could you rename to indicate this is instance state, something like  `processorMetrics` or `pendingMetrics`